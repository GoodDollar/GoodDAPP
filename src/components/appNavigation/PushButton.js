// @flow
import React, { useCallback } from 'react'
import CustomButton, { type ButtonProps } from '../common/buttons/CustomButton'

type PushButtonProps = {
  ...ButtonProps,
  canContinue?: Function,
  params?: any,
  routeName: Route,
  screenProps: { push: (routeName: string, params: any) => void },
  style?: any,
}

/**
 * PushButton
 * This button gets the push action from screenProps. Is meant to be used inside a stackNavigator
 * @param {PushButtonProps} args
 * @param {string} args.routeName
 * @param {object} args.screenProps
 * @param {Function|undefined} args.canContinue
 * @param {object} args.params
 * @param {ButtonProps} props
 */
export const PushButton = ({ routeName, screenProps, canContinue, params, ...props }: PushButtonProps) => {
  const onPress = useCallback(async () => {
    if (screenProps && (await canContinue())) {
      screenProps.push(routeName, params)
    }
  }, [canContinue, screenProps, routeName, params])
  return <CustomButton onPress={onPress} {...props} />
}

PushButton.defaultProps = {
  mode: 'contained',
  dark: true,
  canContinue: () => Promise.resolve(true),
}
