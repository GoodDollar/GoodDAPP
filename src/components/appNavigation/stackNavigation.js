// @flow
import React, { Component, memo, useEffect, useState } from 'react'
import { Platform, SafeAreaView, ScrollView, StyleSheet, View } from 'react-native'
import SideMenu from '@gooddollar/react-native-side-menu'
import { createNavigator, Route, SceneView, SwitchRouter } from '@react-navigation/core'
import { isEqualWith, isFunction, isNumber } from 'lodash'

import { withStyles } from '../../lib/styles'
import { getScreenWidth } from '../../lib/utils/orientation'
import { isWeb } from '../../lib/utils/platform'
import normalize from '../../lib/utils/normalizeText'
import SideMenuPanel from '../sidemenu/SideMenuPanel'
import logger from '../../lib/logger/js-logger'
import CustomButton, { type ButtonProps } from '../common/buttons/CustomButton'
import Blurred from '../common/view/Blurred'
import { GlobalTogglesContext } from '../../lib/contexts/togglesContext'
import BackButtonHandler from './BackButtonHandler'
import NavBar from './NavBar'
import { navigationOptions } from './navigationConfig'
import { PushButton } from './PushButton'

export const DEFAULT_PARAMS = {
  event: undefined,
  receiveLink: undefined,
  reason: undefined,
  code: undefined,
}

const log = logger.child({ from: 'stackNavigation' })

type AppViewProps = {
  descriptors: any,
  navigation: any,
  navigationConfig: any,
  screenProps: any,
}

type AppViewState = {
  stack: Array<any>,
  currentState: any,
}

/**
 * Component wrapping the stack navigator.
 * It holds the pop, push, gotToRoot and goToParent navigation logic and inserts on top the NavBar component.
 * Params are passed as initial state for next screen.
 * This navigation actions are being passed via navigationConfig to children components
 */
class AppView extends Component<AppViewProps, AppViewState> {
  state = {
    stack: [],
    currentState: {},
  }

  /**
   * marks route transistion
   */
  trans: boolean = false

  shouldComponentUpdate(nextProps, nextState) {
    if (this.trans) {
      return false
    }
    const navStateChanged =
      isEqualWith(nextProps, this.props, (val1, val2) => (isFunction(val1) && isFunction(val2) ? true : undefined)) ===
      false
    const stateChanged = isEqualWith(nextState, this.state) === false

    return navStateChanged || stateChanged
  }

  /**
   * handler for back Button on Android
   */
  backButton = null

  componentDidMount() {
    const { toggleClickListener } = this

    this.backButton = new BackButtonHandler({ defaultAction: this.pop })
    toggleClickListener('add')
  }

  componentWillUnmount() {
    const { toggleClickListener } = this

    toggleClickListener('remove')
    this.backButton.unregister()
  }

  toggleClickListener = eventAction => {
    if (!isWeb) {
      return
    }

    document[eventAction + 'EventListener']('mousedown', this.handleClickOutside)
  }

  handleClickOutside = event => {
    const { setMenu, isMenuOpened } = this

    if (isWeb && isMenuOpened && event.target === document.documentElement) {
      setMenu(false)
    }
  }

  /**
   * getComponent gets the component and props and returns the same component except when
   * shouldNavigateToComponent is present in component and not complaining
   * This function can be written in every component that needs to prevent access
   * if there is not in a correct navigation flow.
   * Example: doesn't makes sense to navigate to Amount if there is no nextRoutes
   * @param {React.Component} Component
   * @param props
   */
  getComponent = (descriptor, screenProps) => {
    const Component = descriptor.getComponent()
    const { shouldNavigateToComponent = () => true } = Component

    return shouldNavigateToComponent({ screenProps }) ? Component : ReturnBack
  }

  /**
   * Pops from stack
   * If there is no screen on the stack navigates to initial screen on stack (goToRoot)
   * If we are currently in the first screen go to ths screen that created the stack (goToParent)
   * we can use this to navigate back to previous screen with adding new params
   *
   * @param {object} params new params to add to previous screen screenState
   */
  pop = (params?: any) => {
    const { navigation } = this.props

    const currentParams = navigation.state.routes[navigation.state.index].params
    if (currentParams && currentParams.backPage) {
      this.setState({ currentState: {}, stack: [] }, () => {
        this.trans = false
        navigation.navigate(currentParams.backPage, currentParams.navigationParams)
      })
      return
    }

    const nextRoute = this.state.stack.pop()
    if (nextRoute) {
      this.trans = true
      const navigationParams = nextRoute.state
      this.setState({ currentState: { ...navigationParams, ...params, route: nextRoute.route } }, () => {
        this.trans = false
        navigation.navigate(nextRoute.route, navigationParams)
      })
    } else if (navigation.state.index === 0) {
      this.goToParent()
    } else {
      this.goToRoot()
    }
  }

  /**
   * Push a route to the stack
   * The stack is maintained in stack property to be able to navigate back and forward
   */
  push = (nextRoute, params) => {
    const { navigation } = this.props
    const navigationParams = params || {}
    const route = navigation.state.routes[navigation.state.index].key
    this.trans = true
    this.setState(
      state => {
        return {
          stack: [
            ...state.stack,
            {
              route,
              state: state.currentState,
            },
          ],
          currentState: { ...params, route },
        }
      },
      state => {
        this.trans = false
        navigation.navigate(nextRoute, { ...navigationParams, route })
      },
    )
  }

  /**
   * Navigates to root screen. First on stack
   */
  goToRoot = () => {
    const { navigation } = this.props
    this.trans = true
    this.setState(
      {
        stack: [],
        currentState: {},
      },
      () => {},
    )
    const route = navigation.state.routes[0]
    route.params = {
      ...route.params,
      ...DEFAULT_PARAMS,
    }
    this.trans = false

    // NOTICE: for some reason this doesnt work when inside setState callback only in gotoRoot
    // and when called from a page like SendLinkSummary when opening a payment request, ie not opening dashboard first
    // not sure that we need to keep stack as state variable at all

    navigation.navigate(route)
  }

  /**
   * Navigates to specific screen with custom parameters as query string. and resetting the stack
   */
  navigateTo = (nextRoute: string, params: any) => {
    const { navigation } = this.props
    const route = navigation.state.routes[navigation.state.index].key
    this.trans = true
    this.setState(
      state => {
        return {
          stack: [],
          currentState: { ...params, route },
        }
      },
      () => {
        navigation.navigate(nextRoute)
        this.trans = false
      },
    )
  }

  /**
   * Navigates to the screen that created the stack (backRouteName)
   */
  goToParent = () => {
    const { navigation, navigationConfig } = this.props

    if (navigationConfig.backRouteName) {
      this.trans = true
      this.setState({ currentState: {}, stack: [] }, () => {
        navigation.navigate(navigationConfig.backRouteName)
        this.trans = false
      })
    }
  }

  /**
   * Screen states are being stored by this component
   * This way it can be kept between screens
   */
  setScreenState = data => {
    this.setState(state => ({ currentState: { ...state.currentState, ...data } }))
  }

  render() {
    const { isMenuOn, setMenu } = this.context
    this.setMenu = setMenu
    this.isMenuOpened = isMenuOn
    const { descriptors, navigation, navigationConfig, screenProps: incomingScreenProps } = this.props
    const activeKey = navigation.state.routes[navigation.state.index].key
    const descriptor = descriptors[activeKey]

    const {
      title,
      isBridge,
      navigationBar: NavigationBar,
      navigationBarHidden,
      backButtonHidden,
      disableScroll,
      backToWallet = false,
    } = descriptor.options

    const screenProps = {
      ...incomingScreenProps,
      navigationConfig,
      push: this.push,
      goToRoot: this.goToRoot,
      goToParent: this.goToParent,
      navigateTo: this.navigateTo,
      pop: this.pop,
      screenState: this.state.currentState,
      setScreenState: this.setScreenState,
    }

    log.info('stackNavigation Render: FIXME rerender', descriptor, activeKey, isMenuOn)

    const Component = this.getComponent(descriptor, screenProps)
    const pageTitle = title || activeKey

    return (
      <React.Fragment>
        {isMenuOn && (
          <View style={[styles.sideMenuContainer, styles.menuOpenStyle]} ref={this.wrapperRef}>
            <SideMenu
              menuPosition="right"
              isOpen={true}
              disableGestures={true}
              onChange={this.setMenu}
              menu={
                <SafeAreaView style={styles.safeArea}>
                  <SideMenuPanel navigation={navigation} />
                </SafeAreaView>
              }
            />
          </View>
        )}
        <Blurred ref={this.blurRef} whenSideMenu>
          {!navigationBarHidden &&
            (NavigationBar ? (
              <>
                <NavigationBar title={pageTitle} goBack={backButtonHidden ? undefined : this.pop} />
                {isBridge && (
                  <NavBar
                    backToWallet={backToWallet}
                    goBack={backButtonHidden ? undefined : this.pop}
                    title={pageTitle}
                  />
                )}
              </>
            ) : (
              <NavBar backToWallet={backToWallet} goBack={backButtonHidden ? undefined : this.pop} title={pageTitle} />
            ))}
          {disableScroll ? (
            <SceneView navigation={descriptor.navigation} component={Component} screenProps={screenProps} />
          ) : (
            <ScrollView bounces={false} style={styles.scrollView} contentContainerStyle={styles.scrollableView}>
              <SceneView navigation={descriptor.navigation} component={Component} screenProps={screenProps} />
            </ScrollView>
          )}
        </Blurred>
      </React.Fragment>
    )
  }
}

AppView.contextType = GlobalTogglesContext

const styles = StyleSheet.create({
  scrollView: {
    display: 'flex',
  },
  scrollableView: {
    flexGrow: 1,
    display: 'flex',
  },
  sideMenuContainer: {
    top: 0,
    left: 0,
    bottom: 0,
    right: 0,
    position: 'absolute',
    transform: [
      {
        translateX: Platform.select({
          web: '200vw',
          default: getScreenWidth() * 2,
        }),
      },
    ],
    zIndex: 100,
  },
  menuOpenStyle: {
    transform: [{ translateX: 0 }],
  },
  hideMenu: {
    display: 'none',
  },
  safeArea: {
    padding: 0,
    paddingHorizontal: 0,
    paddingVertical: 0,
    flex: 1,
  },
})

/**
 * Returns a navigator with a navbar wrapping the routes.
 * This function is meant to be used to create a new stack navigation with the given routes.
 * @param {[Route]} routes: Array with routes in the stack
 * @param {Object} navigationConfig
 */
export const createStackNavigator = (routes: any, navigationConfig: any) => {
  const defaultNavigationConfig = {
    backRouteName: 'Home',
  }

  return createNavigator(AppView, SwitchRouter(routes), {
    ...defaultNavigationConfig,
    ...navigationConfig,
    navigationOptions,
  })
}

const ReturnBack = memo(({ screenProps }) => {
  const { goToParent } = screenProps

  useEffect(() => {
    goToParent()
    log.debug('Redirecting to parent:', 'shouldNavigateToComponent() returned false')
  }, [goToParent])

  return null
})

type BackButtonProps = {
  ...ButtonProps,
  routeName?: Route,
  screenProps: { goToParent: () => void },
}

/**
 * BackButton
 * This button gets the goToParent action from screenProps. Is meant to be used inside a stackNavigator
 * @param {ButtonProps} props
 */
const BackButtonComponent = (props: BackButtonProps) => {
  const { disabled, screenProps, children, mode, color, styles, textStyle = {} } = props

  return (
    <CustomButton
      {...props}
      compact={true}
      mode={mode || 'text'}
      color={color || '#A3A3A3'}
      disabled={disabled}
      onPress={screenProps.goToParent}
      textStyle={[styles.cancelButton, textStyle]}
    >
      {children}
    </CustomButton>
  )
}

const getStylesFromProps = ({ theme }) => ({
  cancelButton: {
    color: theme.colors.gray80Percent,
    fontSize: normalize(14),
    fontWeight: '500',
  },
})

export const BackButton = withStyles(getStylesFromProps)(BackButtonComponent)

type NextButtonProps = {
  ...ButtonProps,
  values: {},
  screenProps: { push: (routeName: string, params: any) => void },
  nextRoutes: [string],
  label?: string,
  canContinue?: Function,
}

/**
 * NextButton
 * This button gets the nextRoutes param and creates a Push to the next screen and passes the rest of the array which are
 * next screens for further Components. Is meant to be used inside a stackNavigator
 * @param {any} props
 */
export const NextButton = ({
  disabled,
  values,
  screenProps,
  nextRoutes: nextRoutesParam,
  label,
  canContinue,
  loading,
}: NextButtonProps) => {
  const [next, ...nextRoutes] = nextRoutesParam ? nextRoutesParam : []
  return (
    <PushButton
      mode="contained"
      disabled={disabled || !next}
      screenProps={{ ...screenProps }}
      params={{ ...values, nextRoutes }}
      routeName={next}
      style={Platform.OS === 'web' && { flex: 2 }}
      canContinue={canContinue}
      loading={loading}
    >
      {label || 'Next'}
    </PushButton>
  )
}

type UseScreenProps = { setScreenState?: {}, screenState?: {} }

/**
 * Hook to get screen state from stack or from useState hook if there is no setScreenState function
 */
export const useScreenState = ({ setScreenState, screenState }: UseScreenProps): any => {
  const [state, setState] = useState<any>()

  if (setScreenState) {
    return [screenState || {}, setScreenState]
  }

  return [state || {}, setState]
}

const traverseRoutes = (navState, onLeaf) => {
  const { index, routes } = navState
  let segment = routes[index]

  for (;;) {
    const { index, routes } = segment

    onLeaf(segment)

    if (!isNumber(index)) {
      break
    }

    segment = routes[index]
  }
}

export const getRoutePath = navState => {
  let path = ''

  traverseRoutes(navState, ({ key }) => (path += '/' + key))
  return path
}

export const getRouteName = navState => {
  let name = ''

  traverseRoutes(navState, ({ routeName }) => (name = routeName))
  return name
}
