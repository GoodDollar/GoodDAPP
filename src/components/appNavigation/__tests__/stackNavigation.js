import React from 'react'
import renderer from 'react-test-renderer'
import { PushButton as ImportedPushButton } from '../PushButton'
import { NextButton as ImportedNextButton } from '../stackNavigation'
import { withThemeProvider } from '../../../__tests__/__util__'
const PushButton = withThemeProvider(ImportedPushButton)
const NextButton = withThemeProvider(ImportedNextButton)

// Note: test renderer must be required after react-native.

describe('PushButton', () => {
  it('matches snapshot', async () => {
    let component

    // eslint-disable-next-line require-await
    await renderer.act(async () => (component = renderer.create(<PushButton>Push</PushButton>)))
    const tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })
})

describe('NextButton', () => {
  it('matches snapshot', async () => {
    let component

    // eslint-disable-next-line require-await
    await renderer.act(async () => (component = renderer.create(<NextButton />)))
    const tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })
})
