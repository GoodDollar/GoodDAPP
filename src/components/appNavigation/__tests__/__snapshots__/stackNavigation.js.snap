// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NextButton matches snapshot 1`] = `
<div
  className="css-view-1dbjc4n"
  style={
    Object {
      "WebkitBoxFlex": 2,
      "WebkitBoxPack": "center",
      "WebkitFlexBasis": "0%",
      "WebkitFlexGrow": 2,
      "WebkitFlexShrink": 1,
      "WebkitJustifyContent": "center",
      "backgroundColor": "rgba(0,0,0,0.12)",
      "borderBottomColor": "rgba(0,0,0,0.00)",
      "borderBottomLeftRadius": "4px",
      "borderBottomRightRadius": "4px",
      "borderBottomStyle": "solid",
      "borderBottomWidth": "0px",
      "borderLeftColor": "rgba(0,0,0,0.00)",
      "borderLeftStyle": "solid",
      "borderLeftWidth": "0px",
      "borderRightColor": "rgba(0,0,0,0.00)",
      "borderRightStyle": "solid",
      "borderRightWidth": "0px",
      "borderTopColor": "rgba(0,0,0,0.00)",
      "borderTopLeftRadius": "4px",
      "borderTopRightRadius": "4px",
      "borderTopStyle": "solid",
      "borderTopWidth": "0px",
      "flexBasis": "0%",
      "flexGrow": 2,
      "flexShrink": 1,
      "justifyContent": "center",
      "minWidth": "auto",
      "msFlexNegative": 1,
      "msFlexPack": "center",
      "msFlexPositive": 2,
      "msFlexPreferredSize": "0%",
    }
  }
>
  <div
    aria-disabled={true}
    className="css-view-1dbjc4n"
    disabled={true}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    role="button"
    style={
      Object {
        "borderBottomLeftRadius": "4px",
        "borderBottomRightRadius": "4px",
        "borderTopLeftRadius": "4px",
        "borderTopRightRadius": "4px",
        "overflowX": "hidden",
        "overflowY": "hidden",
        "position": "relative",
      }
    }
  >
    <div
      className="css-view-1dbjc4n"
      style={
        Object {
          "WebkitAlignItems": "center",
          "WebkitBoxAlign": "center",
          "WebkitBoxDirection": "normal",
          "WebkitBoxOrient": "horizontal",
          "WebkitBoxPack": "center",
          "WebkitFlexDirection": "row",
          "WebkitJustifyContent": "center",
          "alignItems": "center",
          "flexDirection": "row",
          "justifyContent": "center",
          "msFlexAlign": "center",
          "msFlexDirection": "row",
          "msFlexPack": "center",
        }
      }
    >
      <div
        className="css-text-901oao css-textOneLine-bfa6kz"
        dir="auto"
        style={
          Object {
            "color": "rgba(0,0,0,0.32)",
            "direction": "ltr",
            "fontFamily": "Roboto-Medium, Roboto, \\"Helvetica Neue\\", Helvetica, Arial, sans-serif",
            "letterSpacing": "1px",
            "marginBottom": "9px",
            "marginLeft": "8px",
            "marginRight": "8px",
            "marginTop": "9px",
            "textAlign": "center",
          }
        }
      >
        <span
          className="css-text-901oao css-textHasAncestor-16my406"
          dir="auto"
          style={
            Object {
              "color": "rgba(255,255,255,1.00)",
              "direction": "ltr",
              "fontFamily": "Roboto, \\"Helvetica Neue\\", Helvetica, Arial, sans-serif",
              "fontSize": "18px",
              "fontWeight": "bold",
              "paddingBottom": "0px",
              "paddingLeft": "0px",
              "paddingRight": "0px",
              "paddingTop": "0px",
              "textAlign": "center",
              "textTransform": "uppercase",
            }
          }
        >
          Next
        </span>
      </div>
    </div>
  </div>
</div>
`;

exports[`PushButton matches snapshot 1`] = `
<div
  className="css-view-1dbjc4n"
  style={
    Object {
      "WebkitBoxPack": "center",
      "WebkitJustifyContent": "center",
      "backgroundColor": "rgba(85,85,85,1.00)",
      "borderBottomColor": "rgba(0,0,0,0.00)",
      "borderBottomLeftRadius": "4px",
      "borderBottomRightRadius": "4px",
      "borderBottomStyle": "solid",
      "borderBottomWidth": "0px",
      "borderLeftColor": "rgba(0,0,0,0.00)",
      "borderLeftStyle": "solid",
      "borderLeftWidth": "0px",
      "borderRightColor": "rgba(0,0,0,0.00)",
      "borderRightStyle": "solid",
      "borderRightWidth": "0px",
      "borderTopColor": "rgba(0,0,0,0.00)",
      "borderTopLeftRadius": "4px",
      "borderTopRightRadius": "4px",
      "borderTopStyle": "solid",
      "borderTopWidth": "0px",
      "boxShadow": "0px 0.75px 1.5px rgba(0,0,0,0.24)",
      "justifyContent": "center",
      "minWidth": "auto",
      "msFlexPack": "center",
    }
  }
>
  <div
    className="css-cursor-18t94o4 css-view-1dbjc4n"
    data-focusable={true}
    disabled={false}
    onKeyDown={[Function]}
    onKeyPress={[Function]}
    onKeyUp={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    role="button"
    style={
      Object {
        "borderBottomLeftRadius": "4px",
        "borderBottomRightRadius": "4px",
        "borderTopLeftRadius": "4px",
        "borderTopRightRadius": "4px",
        "cursor": "pointer",
        "msTouchAction": "manipulation",
        "overflowX": "hidden",
        "overflowY": "hidden",
        "position": "relative",
        "touchAction": "manipulation",
      }
    }
    tabIndex="0"
  >
    <div
      className="css-view-1dbjc4n"
      style={
        Object {
          "WebkitAlignItems": "center",
          "WebkitBoxAlign": "center",
          "WebkitBoxDirection": "normal",
          "WebkitBoxOrient": "horizontal",
          "WebkitBoxPack": "center",
          "WebkitFlexDirection": "row",
          "WebkitJustifyContent": "center",
          "alignItems": "center",
          "flexDirection": "row",
          "justifyContent": "center",
          "msFlexAlign": "center",
          "msFlexDirection": "row",
          "msFlexPack": "center",
        }
      }
    >
      <div
        className="css-text-901oao css-textOneLine-bfa6kz"
        dir="auto"
        style={
          Object {
            "color": "rgba(255,255,255,1.00)",
            "direction": "ltr",
            "fontFamily": "Roboto-Medium, Roboto, \\"Helvetica Neue\\", Helvetica, Arial, sans-serif",
            "letterSpacing": "1px",
            "marginBottom": "9px",
            "marginLeft": "8px",
            "marginRight": "8px",
            "marginTop": "9px",
            "textAlign": "center",
          }
        }
      >
        <span
          className="css-text-901oao css-textHasAncestor-16my406"
          dir="auto"
          style={
            Object {
              "color": "rgba(255,255,255,1.00)",
              "direction": "ltr",
              "fontFamily": "Roboto, \\"Helvetica Neue\\", Helvetica, Arial, sans-serif",
              "fontSize": "18px",
              "fontWeight": "bold",
              "paddingBottom": "0px",
              "paddingLeft": "0px",
              "paddingRight": "0px",
              "paddingTop": "0px",
              "textAlign": "center",
              "textTransform": "uppercase",
            }
          }
        >
          Push
        </span>
      </div>
    </div>
  </div>
</div>
`;
